java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java

```diff
@@ -0,0 +1,248 @@
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.npe;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Lists;
+import org.sonar.java.resolve.JavaSymbol;
+import org.sonar.java.symexecengine.DataFlowVisitor;
+import org.sonar.java.symexecengine.State;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ArrayAccessExpressionTree;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+public class NpeVisitor extends DataFlowVisitor {
+
+  public NpeVisitor(List<VariableTree> parameters) {
+    for (VariableTree parameter : parameters) {
+      super.visitVariable(parameter);
+      State state;
+      if (parameter.symbol().metadata().isAnnotatedWith("javax.annotation.CheckForNull") || parameter.symbol().metadata().isAnnotatedWith("javax.annotation.Nullable")) {
+        state = new NPEState.Null(parameter);
+      } else {
+        state = new NPEState.NotNull(parameter);
+      }
+      executionState.markValueAs(parameter.symbol(), state);
+    }
+  }
+
+  @Override
+  protected boolean isSymbolRelevant(Symbol symbol) {
+    return true;
+  }
+
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    super.visitAssignmentExpression(tree);
+    Symbol symbol = getSymbol(tree.variable());
+    if (symbol != null && symbol.owner().isMethodSymbol()) {
+      setStateOfValue(symbol, tree.expression());
+    }
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    super.visitVariable(tree);
+    setStateOfValue(tree.symbol(), tree.initializer());
+  }
+
+  @Override
+  public void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+    super.visitArrayAccessExpression(tree);
+    if (!tree.expression().is(Tree.Kind.MEMBER_SELECT) && isNullTree(tree.expression())) {
+      executionState.reportIssue(tree);
+    }
+  }
+
+  private void setStateOfValue(Symbol symbol, @Nullable ExpressionTree tree) {
+    if (tree == null || isNullTree(tree)) {
+      executionState.markValueAs(symbol, new NPEState.Null(tree));
+    } else {
+      executionState.markValueAs(symbol, new NPEState.NotNull(tree));
+    }
+  }
+
+  private boolean isNullTree(ExpressionTree tree) {
+    Symbol symbol = getSymbol(tree);
+    return (symbol != null && executionState.hasState(symbol, NPEState.Null.class)) ||
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)